docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,364 @@
+
+> Note: Dependent ordering among tasks will be addresed later which may enable concurrent execution of tasks.
+
+#### Deletion
+
+`Etcd-druid` will ensure that it garbage collects the custom resources and any other k8s resources created to realise the task as per `.spec.ttlSecondsAfterFinished` if defined in the [spec](#spec) or it assumes a default expiry time.
+
+### Use Cases
+
+#### Recovery from permanent quorum loss
+
+Currently identification and recovery from permament quorum loss is done manaually. The current proposal keeps the identification as manual, requiring an human operator to intervene and confirm that there is indeed a permanent quorum loss with no auto healing possibility. Once that is established then the next step to recover the cluster is now proposed to be automated. Recovery today is a [multi-step process](https://github.com/gardener/etcd-druid/blob/master/docs/operations/recovery-from-permanent-quorum-loss-in-etcd-cluster.md) and needs to be done carefully. Automation would ensure that we eliminate any errors from an operator.
+
+##### Task Config
+
+We do not need any config for this task. When creating an instance of `EtcdOperatorTask` for this scenario, `.spec.config` will be set to nil (unset).
+
+##### Pre-Conditions
+
+* There should be a quorum loss in a multi-member etcd cluster. For a single-member etcd cluster, invoking this task is unnecessary as the restoration of the single member is automatically handled by the backup-restore process.
+* There should not already be a permanent-quorum-loss-recovery-task running for the same etcd cluster.
+
+#### Trigger on-demand snapshot compaction
+
+`etcd-druid` provides a configurable [etcd-events-threshold](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/02-snapshot-compaction.md#druid-flags). If and when this threshold is breached then only a [snapshot compaction](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/02-snapshot-compaction.md) is triggered for etcd cluster. However, there are scenarios where an ad-hoc snapshot compaction is required.
+
+##### Possible scenarios
+
+* Full snapshot taken via compaction job is better as compared to full snapshot taken directly from etcd cluster as full snapshot taken via snapshot compaction is taken after [defragmentation](https://etcd.io/docs/v3.2/op-guide/maintenance/#defragmentation) which makes it smaller (more compacted) in size.
+* Full snapshot triggered on etcd cluster might cause load on running etcd cluster but full snapshot taken by compaction job is developed separately with zero effect on the running cluster.
+
+##### Task Config
+
+We do not need any config for this task. When creating an instance of `EtcdOperatorTask` for this scenario, `.spec.config` will be set to nil (unset).
+
+##### Pre-Conditions
+
+* There should not be a `on-demand snapshot-compaction` task already running for the same etcd cluster.
+
```

Comment:
Semantically you are right. You can only order the tasks if there are more than one queued tasks at the time you decide to pick up one for processing. If you have picked up one for processing and then the other task comes in which ideally should be executed first (only for optimisation reason) then the pre-condition would pass for the in-flight task because the other task got queued only after executing the pre-condition checks which would happen in micro/milli seconds.

So if you are trying to use a priority queue then you must assign relative priorities to tasks. This is different from a pre-condition which is only evaluated when a task is picked up for execution. It is a guard to allow or reject a task.

Reply:
From what I understand, the proposal currently only proposes a FIFO method of acting on existing tasks and states dependent ordering will only be addressed later, so this would be out of scope of this proposal currently, unless we plan on expanding the proposal to define these as well now. This comes under dependent ordering, I presume?

Reply:
> if you are trying to use a priority queue then you must assign relative priorities to tasks.

Yeah I meant to give priority to tasks but after looking at the response from Ishan about why ordering is not that helpful and not necessary at times, this can be parked for later if at all the need arises. Thanks